examples/dft.masm

```diff
@@ -0,0 +1,288 @@
+                    push.3                     # stack: [3]
+                else
+                    dup                        # stack: [n, n]
+                    eq.16                      # stack: [0 or 1, n]
+                    if.true
+                        # n == 16
+                        drop                       # stack: []
+                        push.4                     # stack: [4]
+                    else
+                        # n unsupported; halt
+                        push.1
+                        assertz
+                    end
+                end
+            end
+        end
+
+    end
+end
+
+#! Computes the primitive root of unity for a subgroup of size 2^n.
+#!
+#! Input: [log2_n], log2 of the size of the input list (must be <= 32)
+#! Output: [root_of_unity], where `root_of_unity` is the primitive root of unity of order 2^n
+#!
+#! Equivalent to Winterfell's implementation: 
+#! https://github.com/facebook/winterfell/blob/4543689f73a2efb9d30927535de7b4efe7e1802d/math/src/field/traits.rs#L254-L259
+proc.get_root_of_unity
+    push.TWO_ADIC_ROOT_OF_UNITY.TWO_ADICITY     # stack: [TWO_ADICITY, TWO_ADIC_ROOT_OF_UNITY, log2_n]
+    movup.2                                     # stack: [log2_n, TWO_ADICITY, TWO_ADIC_ROOT_OF_UNITY]
+    sub                                         # stack: [TWO_ADICITY - log2_n, TWO_ADIC_ROOT_OF_UNITY]
+    pow2                                        # stack: [2^(TWO_ADICITY - log2_n), TWO_ADIC_ROOT_OF_UNITY]
+    exp                                         # stack: [TWO_ADIC_ROOT_OF_UNITY^(2^(TWO_ADICITY - log2_n))]
+end
+
+#! Computes the kth element of the "frequency domain" (i.e. of the transformed list).
+#!
+#! Input: [k, root_of_unity, n, v_0, ..., v_{n-1}, ...]
+#! Output: [f_k, ...]
+#!
+#! Locals
+#! 0: k
+#! 1: root_of_unity
+#! 2: n (length of the list)
+#! 3: j (counter)
+#! 4: result (partial result of the computation)
+proc.f_k.5
+    loc_store.0                                # store k
+    loc_store.1                                # store root_of_unity
+    loc_store.2                                # store n
+    push.0                                     # Prepare j = 0
+    loc_store.3                                # Set j = 0
+    push.0                                     # Prepare result = 0
+    loc_store.4                                # Set result = 0
+
+    # Stack here: [v0, ..., v{n-1}, ...]
+
+    push.1                                     # push 1 to enter the loop
+    while.true
+        # compute `root_of_unity^(jk)`
+        loc_load.1                             # stack: [root_of_unity, v{j}, ..., v{n-1}]
+        loc_load.0                             # stack: [k , root_of_unity, v{j}, ..., v{n-1}]
+        loc_load.3                             # stack: [j, k , root_of_unity, v{j}, ..., v{n-1}]
+        mul                                    # stack: [j*k , root_of_unity, v{j}, ..., v{n-1}]
```

Comment:
only `j` changes with every iteration, right? Can you load the other values out of the loop and the multiply with what you need?

Reply:
I believe [this comment](https://github.com/0xPolygonMiden/examples/pull/150#discussion_r1366623645) invalidates this one?

Rereading this now, I used memory to preserve on every iteration `root_of_unity`, `k` and `j`. I believe the canonical way would have been to move those values up/down with `movup`/`movdn`, as it results in fewer cycles in most cases (as one `loc_load`/`loc_store` pair costs 6-8 cycles).